Add BERT Loses Patience (Patience-based Early Exit)

JetRunner opened the pull request:
Add BERT Loses Patience (Patience-based Early Exit) based on the paper https://arxiv.org/abs/2006.04152 and the official implementation https://github.com/JetRunner/PABEE

It's impossible to make PABEE's ALBERT and BERT compatible with the standard API (e.g., `run_glue.py` so I keep the modeling files in a separate directory, under `example/bert_loses_patience`, instead of putting them alongside `modeling_bert.py`)

sshleifer commented on examples/bert-loses-patience/pabee/modeling_pabee_albert.py:
```diff
@@ -0,0 +1,990 @@
+# coding=utf-8
+# Copyright 2018 Google AI, Google Brain, the HuggingFace Inc. team and Microsoft Corporation.
```
year seems wrong,

sshleifer commented on examples/bert-loses-patience/pabee/modeling_pabee_albert.py:
```diff
@@ -0,0 +1,990 @@
+logger = logging.getLogger(__name__)
+
+
+ALBERT_PRETRAINED_MODEL_ARCHIVE_LIST = [
```
do you want to import the things that are identical from modeling_albert.py? I would find that easier to read, but don't feel too strongly.

sshleifer commented on examples/bert-loses-patience/pabee/modeling_pabee_albert.py:
```diff
@@ -0,0 +1,990 @@
+    def set_regression_threshold(self, threshold):
+        self.regression_threshold = threshold
+
+    def set_patience(self, patience):
+        self.patience = patience
+
+    def reset_stats(self):
+        self.inference_instances_num = 0
+        self.inference_layers_num = 0
+
+    def log_stats(self):
+        avg_inf_layers = self.inference_layers_num / self.inference_instances_num
+        message = f"*** Patience = {self.patience} Avg. Inference Layers = {avg_inf_layers:.2f} Speed Up = {1 - avg_inf_layers / self.config.num_hidden_layers:.2f} ***"
+        print(message)
```
Another design could be
```python
class AlbertModelWithPabee(AlbertModel):
    def set_patience
	def other_new_methods...
```

sshleifer commented on examples/bert-loses-patience/run_glue_with_pabee.py:
```diff
@@ -0,0 +1,705 @@
+        torch.cuda.manual_seed_all(args.seed)
+
+
+def train(args, train_dataset, model, tokenizer):
```
Would be interesting to see if this can work with `Trainer`

JetRunner replied:
It'll be great. But we have to be cautious here not to change a lot :(

sshleifer replied:
Best way to do that IMO is to write an integration test that asserts that the loss after 1 hardcoded batch is as you expect, then refactor without breaking the test.

JetRunner commented:
Hi @LysandreJik , my worry is that PABEE is not a standard inference method. It cannot deal with batch inference. Also, it can only support classification & regression (no tagging, no summarization, etc.). 

From another aspect, as a researcher, when I try to do a little tweak with ALBERT, I won't be happy if someone adds some new stuff into the model and it'll add unnecessary burdens to the researchers. They'll definitely hate it. So IMO, it's better to implement them separately. Also, as @sshleifer suggested, I refactored the model to inherit `AlbertTransformer` and `AlbertModel`, etc.

sshleifer commented on examples/bert-loses-patience/README.md:
```diff
@@ -0,0 +1,92 @@
+  --data_dir "$GLUE_DIR/$TASK_NAME" \
+  --max_seq_length 128 \
+  --per_gpu_train_batch_size 32 \
+  --per_gpu_eval_batch_size 32 \
```
why is there an `eval_batch_size` clarg if you can't support batched inference?

JetRunner replied:
I'll add a check

sshleifer commented on examples/bert-loses-patience/README.md:
```diff
@@ -0,0 +1,92 @@
+  --per_gpu_train_batch_size 32 \
+  --per_gpu_eval_batch_size 32 \
+  --learning_rate 2e-5 \
+  --save_steps 50 \
```
A good name for save_steps is `val_check_interval`, if I understand it properly.

JetRunner replied:
No, this one is for saving the checkpoints

sshleifer commented on examples/bert-loses-patience/README.md:
```diff
@@ -0,0 +1,92 @@
+export TASK_NAME=MRPC
+
+python ./run_glue_with_pabee.py \
+  --model_type bert/albert \
```
just pick one, so that people can copy paste. This code should actually work.

sshleifer commented on examples/bert-loses-patience/README.md:
```diff
@@ -0,0 +1,92 @@
+  --do_lower_case \
+  --data_dir "$GLUE_DIR/$TASK_NAME" \
+  --max_seq_length 128 \
+  --per_gpu_train_batch_size 32 \
```
do we need this clarg for inference?

JetRunner replied:
I'll remove the unused ones

sshleifer commented on examples/bert-loses-patience/pabee/modeling_pabee_albert.py:
```diff
@@ -0,0 +1,331 @@
+            pooled_output = self.pooler_activation(self.pooler(encoder_outputs[0][:, 0]))
+            res = [output_layers[self.config.num_hidden_layers - 1](pooled_output)]
+        else:
+            patient_counter = 0
+            patient_result = None
+            calculated_layer_num = 0
+            for i in range(self.config.num_hidden_layers):
+                calculated_layer_num += 1
+                encoder_outputs = self.encoder.adaptive_forward(
+                    encoder_outputs, current_layer=i, attention_mask=extended_attention_mask, head_mask=head_mask,
+                )
+
+                pooled_output = self.pooler_activation(self.pooler(encoder_outputs[0][:, 0]))
+                logits = output_layers[i](pooled_output)
+                if regression:
+                    labels = logits.detach()
+                    if patient_result is not None:
+                        patient_labels = patient_result.detach()
+                    if (patient_result is not None) and torch.abs(patient_result - labels) < self.regression_threshold:
+                        patient_counter += 1
+                    else:
+                        patient_counter = 0
+                else:
+                    labels = logits.detach().argmax(dim=1)
+                    if patient_result is not None:
+                        patient_labels = patient_result.detach().argmax(dim=1)
+                    if (patient_result is not None) and torch.all(labels.eq(patient_labels)):
+                        patient_counter += 1
+                    else:
+                        patient_counter = 0
+
+                patient_result = logits
+                if patient_counter == self.patience:
+                    break
+            res = [patient_result]
+            self.inference_layers_num += calculated_layer_num
+            self.inference_instances_num += 1
```
can we better encapsulate this in a method, like `_adaptive_forward`. I think @LysandreJik is OK with that as long as the method is not user facing.

JetRunner replied:
I don't think it makes sense to put these into another method. These codes are used only once. Also, I think it should be contained in `forward`, otherwise many parameters have to be added to the encapsulated function.

sshleifer commented on examples/bert-loses-patience/pabee/modeling_pabee_bert.py:
```diff
@@ -0,0 +1,460 @@
+        return hidden_states
+
+
+class BertPreTrainedModelWithPabee(BertPreTrainedModel):
```
you don't need this, right?

sshleifer commented on examples/bert-loses-patience/pabee/modeling_pabee_bert.py:
```diff
@@ -0,0 +1,460 @@
+    """
+
+
+BERT_START_DOCSTRING = r"""
```
you can import all this stuff I think.

sshleifer commented on examples/bert-loses-patience/pabee/modeling_pabee_bert.py:
```diff
@@ -0,0 +1,460 @@
+        if token_type_ids is None:
+            token_type_ids = torch.zeros(input_shape, dtype=torch.long, device=device)
+
+        # We can provide a self-attention mask of dimensions [batch_size, from_seq_length, to_seq_length]
+        # ourselves in which case we just need to make it broadcastable to all heads.
+        if attention_mask.dim() == 3:
+            extended_attention_mask = attention_mask[:, None, :, :]
+        elif attention_mask.dim() == 2:
+            # Provided a padding mask of dimensions [batch_size, seq_length]
+            # - if the model is a decoder, apply a causal mask in addition to the padding mask
+            # - if the model is an encoder, make the mask broadcastable to [batch_size, num_heads, seq_length, seq_length]
+            if self.config.is_decoder:
+                batch_size, seq_length = input_shape
+                seq_ids = torch.arange(seq_length, device=device)
+                causal_mask = seq_ids[None, None, :].repeat(batch_size, seq_length, 1) <= seq_ids[None, :, None]
+                causal_mask = causal_mask.to(
+                    attention_mask.dtype
+                )  # causal and attention masks must have same type with pytorch version < 1.3
+                extended_attention_mask = causal_mask[:, None, :, :] * attention_mask[:, None, None, :]
+            else:
+                extended_attention_mask = attention_mask[:, None, None, :]
+        else:
+            raise ValueError(
+                "Wrong shape for input_ids (shape {}) or attention_mask (shape {})".format(
+                    input_shape, attention_mask.shape
+                )
+            )
+
+        # Since attention_mask is 1.0 for positions we want to attend and 0.0 for
+        # masked positions, this operation will create a tensor which is 0.0 for
+        # positions we want to attend and -10000.0 for masked positions.
+        # Since we are adding it to the raw scores before the softmax, this is
+        # effectively the same as removing these entirely.
+        extended_attention_mask = extended_attention_mask.to(dtype=next(self.parameters()).dtype)  # fp16 compatibility
+        extended_attention_mask = (1.0 - extended_attention_mask) * -10000.0
+
```
```suggestion
extended_attention_mask: torch.Tensor = self.get_extended_attention_mask(attention_mask, input_shape, device)

```

sshleifer replied:
this applies to Albert also

JetRunner replied:
This seems to be good. But I just copied it from the original bert/albert code

sshleifer replied:
those got cleaned up since you copied I think.

JetRunner replied:
Oh I didn’t realize that. Will sync it up with the latest code.

sshleifer commented on examples/bert-loses-patience/pabee/modeling_pabee_bert.py:
```diff
@@ -0,0 +1,460 @@
+        # attention_probs has shape bsz x n_heads x N x N
+        # input head_mask has shape [num_heads] or [num_hidden_layers x num_heads]
+        # and head_mask is converted to shape [num_hidden_layers x batch x num_heads x seq_length x seq_length]
+        if head_mask is not None:
+            if head_mask.dim() == 1:
+                head_mask = head_mask.unsqueeze(0).unsqueeze(0).unsqueeze(-1).unsqueeze(-1)
+                head_mask = head_mask.expand(self.config.num_hidden_layers, -1, -1, -1, -1)
+            elif head_mask.dim() == 2:
+                head_mask = (
+                    head_mask.unsqueeze(1).unsqueeze(-1).unsqueeze(-1)
+                )  # We can specify head_mask for each layer
+            head_mask = head_mask.to(
+                dtype=next(self.parameters()).dtype
+            )  # switch to fload if need + fp16 compatibility
+        else:
+            head_mask = [None] * self.config.num_hidden_layers
+
```
```suggestion
head_mask = self.get_head_mask(head_mask, self.config.num_hidden_layers)

```

JetRunner commented:
> Hi @LysandreJik , my worry is that PABEE is not a standard inference method. It cannot deal with batch inference. Also, it can only support classification & regression (no tagging, no summarization, etc.).
> 
> From another aspect, as a researcher, when I try to do a little tweak with ALBERT, I won't be happy if someone adds some new stuff into the model and it'll add unnecessary burdens to the researchers. They'll definitely hate it. So IMO, it's better to implement them separately. Also, as @sshleifer suggested, I refactored the model to inherit `AlbertTransformer` and `AlbertModel`, etc.

Since I refactored the code with inherit, I figure it is okay to use `adaptive_forward` since I won't have to overwrite the standard `forward`. Also, it's better to preserve the standard `forward` so we can easily compare `adaptive_forward` to the standard `forward`.

@sshleifer I copy most stuff from the original ALBERT & BERT modeling code so I think it also does not make sense if I refactor the parts I copied.

sshleifer commented:
The code you copied has gotten cleaned up since you copied it, hence the suggestions.

LysandreJik commented on examples/bert-loses-patience/pabee/modeling_pabee_albert.py:
```diff
@@ -0,0 +1,331 @@
+
+        Examples::
+
+            from transformers import AlbertTokenizer, AlbertForSequenceClassification
```
Should be `AlbertForSequenceClassificationWithPabee`

LysandreJik replied:
Like all the other docstrings :smile:

JetRunner replied:
Yeah. I’ll check the docstrings again

LysandreJik commented on examples/bert-loses-patience/pabee/modeling_pabee_albert.py:
```diff
@@ -0,0 +1,331 @@
+
+
+@add_start_docstrings(
+    """Albert Model transformer with a sequence classification/regression head on top (a linear layer on top of
```
Should specify it's an ALBERT model with Pabee?

LysandreJik commented on examples/bert-loses-patience/pabee/modeling_pabee_albert.py:
```diff
@@ -0,0 +1,331 @@
+
+
+@add_start_docstrings(
+    "The bare ALBERT Model transformer outputting raw hidden-states without any specific head on top.",
```
with PABEE

LysandreJik commented on examples/bert-loses-patience/pabee/modeling_pabee_albert.py:
```diff
@@ -0,0 +1,331 @@
+    the pooled output) e.g. for GLUE tasks. """,
+    ALBERT_START_DOCSTRING,
+)
+class AlbertForSequenceClassificationWithPabee(AlbertForSequenceClassification):
```
I'm not sure it's very interesting to override `AlbertForSequenceClassification` for only two items: the `self.num_labels` and `self.dropout`. I would just put `AlbertPreTrainedModel` here, this would prevent initializing a dropout layer + a classifier layer that are not used.

JetRunner replied:
Great point!

LysandreJik commented on examples/bert-loses-patience/pabee/modeling_pabee_albert.py:
```diff
@@ -0,0 +1,331 @@
+                total_weights += ix + 1
+            outputs = (total_loss / total_weights,) + outputs
+
+        return outputs  # (loss), logits, (hidden_states), (attentions)
```
I don't think this can return the hidden states and attentions? cf. `outputs = (logits[-1],)`

JetRunner replied:
Oh that’s the stale comment. Great catch!

LysandreJik commented on examples/bert-loses-patience/pabee/modeling_pabee_bert.py:
```diff
@@ -0,0 +1,403 @@
+logger = logging.getLogger(__name__)
+
+
+class BertEncoderWithPabee(BertEncoder):
+    def adaptive_forward(self, hidden_states, current_layer, attention_mask=None, head_mask=None):
+        layer_outputs = self.layer[current_layer](hidden_states, attention_mask, head_mask[current_layer])
+
+        hidden_states = layer_outputs[0]
+
+        return hidden_states
```
Very clean!

LysandreJik commented on examples/bert-loses-patience/pabee/modeling_pabee_bert.py:
```diff
@@ -0,0 +1,403 @@
+    the pooled output) e.g. for GLUE tasks. """,
+    BERT_START_DOCSTRING,
+)
+class BertForSequenceClassification(BertForSequenceClassification):
+    def __init__(self, config):
+        super().__init__(config)
+
+        self.bert = BertModelWithPabee(config)
+        self.classifiers = nn.ModuleList(
+            [nn.Linear(config.hidden_size, self.config.num_labels) for _ in range(config.num_hidden_layers)]
+        )
+
+        self.init_weights()
```
Same remarks as for `AlbertForSequenceClassification`

JetRunner commented:
> LGTM pending suggestions, test.

Okay, I’ll add a test
